### DISCUSSION TITLE
How to make a window using the WinAPI?

### DISCUSSION THREAD

**Discussion by DreepyYunky** [odin-lang/Odin#3070] [Q&A] [answered]
Any code example out there on the WinAPI for Odin?

**Comment by leidegre** [comment]
Oh I just recently picked up Odin but my understanding is that there's nothing new here. 

```odin
package main

import "core:sys/windows"
import "core:unicode/utf16"

// A Windows specific window
WindowsWindow :: struct {
	hwnd: windows.HWND,
}

main :: proc() {
	// zero initialized memory on stack
	tmp: [64]u16

	// convert to utf-16
	utf16.encode_string(tmp[:], "MyWindowClass")

	cls: windows.WNDCLASSW
	cls.lpszClassName = raw_data(tmp[:]) // raw_data returns the underlying data of a built-in data type
	windows.RegisterClassW(&cls)
}
```

Odin just let's us call directly into the Win32 API which I think is nice. Nothing new to learn.

You could also use glfw or raylib if you wanted to.

**Reply by DreepyYunky** [reply]
This isn't a window example.

Or I might just use SDL2

**Reply by flysand7** [reply]
Yeah that is the example of calling a windows API function from Odin, in particular registering a window class, which is the first step of creating a window. I haven't dealt with windows in a long while so can't help you with the example unfortunately.

I think it will be helpful to look at [winapi docs](https://learn.microsoft.com/en-us/windows/win32/learnwin32/creating-a-window) and try to convert those to Odin (relevant WinAPI functions and constants should be located in `sys/windows`.

Using SDL2 is not a bad idea if you're going to be software-rendering.

**Reply by DreepyYunky** [reply, marked as answer]
After some errors, checking `sys/windows`'s source, and consulting code examples written in C, I did get a window to work.

Maybe I'll create a gist or something like that so others can easily get started

**Reply by leidegre** [reply]
Please do but I think this is one of the great things about Odin that it doesn't ask you to learn anything new, assuming you already know the Win32 API. If you don't well, then there's that but every other window implementation is using that and these APIs (for better, or worse) has been around since Windows 1.0.

**Reply by Hyrtwol** [reply]
Hey I added a PR with a win32 demo that creates a windows [game of life using win32](https://github.com/odin-lang/examples/pull/49)
